File: src/json/multitouch_vi_mode.json.rb
#!/usr/bin/env ruby

# You can generate json by executing the following command on Terminal.
#
# $ ruby ./multitouch_vi_mode.json.rb
#

require 'json'
require_relative '../lib/karabiner.rb'

def main
  #
  # conditions
  #

  condition1 = Karabiner.variable_if('multitouch_extension_finger_count_total', 1)
  condition2 = Karabiner.variable_unless('multitouch_extension_finger_count_total', 0)

  #
  # manipulators
  #

  manipulators = []

  [
    #
    # Finger count == 1
    #

    {
      :from => { 'key_code' => 'k', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'up_arrow' },
      :conditions => [condition1],
    },
    {
      :from => { 'key_code' => 'j', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'down_arrow' },
      :conditions => [condition1],
    },
    {
      :from => { 'key_code' => 'h', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'left_arrow' },
      :conditions => [condition1],
    },
    {
      :from => { 'key_code' => 'l', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'right_arrow' },
      :conditions => [condition1],
    },

    #
    # Finger count >= 2
    #

    {
      :from => { 'key_code' => 'k', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'page_up' },
      :conditions => [condition2],
    },
    {
      :from => { 'key_code' => 'j', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'page_down' },
      :conditions => [condition2],
    },
    {
      :from => { 'key_code' => 'h', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'home' },
      :conditions => [condition2],
    },
    {
      :from => { 'key_code' => 'l', 'modifiers' => Karabiner.from_modifiers },
      :to => { 'key_code' => 'end' },
      :conditions => [condition2],
    },
  ].each do |d|
    manipulators << {
      'type' => 'basic',
      'from' => d[:from],
      'to' => [
        d[:to],
      ],
      'conditions' => d[:conditions],
    }
  end

  puts JSON.pretty_generate(
    'title' => 'Multitouch Vi Mode (rev 2)',
    'maintainers' => ['tekezo'],
    'rules' => [
      'description' => 'Multitouch Vi Mode (rev 2)',
      'available_since' => '12.6.9',
      'manipulators' => manipulators,
    ]
  )
end

main

File: src/json/multitouch_vi_mode.json.js
// JavaScript should be written in ECMAScript 5.1.

function main() {
  console.log(
    JSON.stringify(
      {
        title: 'Multitouch Vi Mode (rev 2)',
        maintainers: ['tekezo'],
        rules: [
          {
            description: 'Multitouch Vi Mode (rev 2)',
            available_since: '12.6.9',
            manipulators: manipulators(),
          },
        ],
      },
      null,
      '  '
    )
  )
}

function manipulators() {
  //
  // conditions
  //

  total1 = {
    type: 'variable_if',
    name: 'multitouch_extension_finger_count_total',
    value: 1,
  }
  total2 = {
    type: 'variable_unless',
    name: 'multitouch_extension_finger_count_total',
    value: 0,
  }

  //
  // manipulators
  //

  const definitions = [
    //
    // Finger count == 1
    //

    {
      from: { key_code: 'k', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'up_arrow' }],
      conditions: [total1],
    },
    {
      from: { key_code: 'j', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'down_arrow' }],
      conditions: [total1],
    },
    {
      from: { key_code: 'h', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'left_arrow' }],
      conditions: [total1],
    },
    {
      from: { key_code: 'l', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'right_arrow' }],
      conditions: [total1],
    },

    //
    // Finger count >= 2
    //

    {
      from: { key_code: 'k', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'page_up' }],
      conditions: [total2],
    },
    {
      from: { key_code: 'j', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'page_down' }],
      conditions: [total2],
    },
    {
      from: { key_code: 'h', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'home' }],
      conditions: [total2],
    },
    {
      from: { key_code: 'l', modifiers: { optional: ['any'] } },
      to: [{ key_code: 'end' }],
      conditions: [total2],
    },
  ]

  const result = []

  definitions.forEach(function (def) {
    result.push({
      type: 'basic',
      from: def.from,
      to: def.to,
      conditions: def.conditions,
    })
  })

  return result
}

main()
